### PR TITLE
Feature/deflated mg

### PR DESCRIPTION
**Pull request by cpviolator**
A stable Thick Restarted Lanczos, with preliminary support for deflated multigrid.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -413,23 +418,30 @@ endif(QUDA_MAGMA)
 if(QUDA_ARPACK)
   add_definitions(-DARPACK_LIB)
   find_package(PkgConfig REQUIRED)
-  if(QUDA_QMP OR QUDA_MPI)
+
```
Comment (maddyscientist) [inline CMakeLists.txt]: Can you explain why the changes to CMake here with respect to `ARPACK`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -589,7 +601,7 @@ endif(QUDA_COVDEV)
 set(CMAKE_F_FLAGS -std=c99 CACHE STRING "Fortran Flags")
 
 # derive whether we need to build the fortran interface
-if(QUDA_INTERFACE_TIFR OR QUDA_INTERFACE_BQCD)
+if(QUDA_INTERFACE_TIFR OR QUDA_INTERFACE_BQCD OR QUDA_ARPACK)
```
Comment (maddyscientist) [inline CMakeLists.txt]: Why does `QUDA_ARPACK` being enabled mean we build QUDA's Fortran interface?

File: include/eigensolve_quda.h [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    double *residua;
+
+    // Device side vector workspace
+    std::vector<ColorSpinorField *> r;
+    std::vector<ColorSpinorField *> d_vecs_tmp;
+
+    Complex *Qmat;
+
+public:
+    EigenSolver(QudaEigParam *eig_param, TimeProfile &profile);
```
Comment (maddyscientist) [inline include/eigensolve_quda.h]: doxygen needed for the constructor and other member functions

File: include/enum_quda.h [hunk 4]
```diff
@@ -129,11 +129,26 @@ extern "C" {
   } QudaInverterType;
 
   typedef enum QudaEigType_s {
-    QUDA_LANCZOS, //Normal Lanczos eigen solver
-    QUDA_IMP_RST_LANCZOS, //implicit restarted lanczos solver
-    QUDA_INVALID_TYPE = QUDA_INVALID_ENUM
+    QUDA_LANCZOS,           // Lanczos solver
```
Comment (maddyscientist) [inline include/enum_quda.h]: This is probably a good chance to clean up the enums here.  Perhaps `QUDA_EIG_LANCZOS`, QUDA_EIG_ARNOLDI`, etc. here.

Also, I'd prefer `QUDA_SPECTRUM_SR_EIG` over `QUDA_SR_EIG_SPRECTRUM`.  Not all enums follow this convention of increasing specificity as we move from left to right, but that was a poor early design choice.  Since this is a new type, let's get this one correct 😉

File: include/enum_quda_fortran.h [hunk 5]
```diff
@@ -126,9 +126,12 @@
...
 
 #define QudaEigType integer(4)
-#define QUDA_LANCZOS 0 //Normal Lanczos eigen solver
-#define QUDA_IMP_RST_LANCZOS 1 //implicit restarted lanczos solver
-#define QUDA_INVALID_TYPE QUDA_INVALID_ENUM
+#define QUDA_LANCZOS 0           // Lanczos solver
+#define QUDA_IMP_RST_LANCZOS 1   // Implicitly restarted lanczos solver
+#define QUDA_THICK_RST_LANCZOS 2 // Thick restarted lanczos solver
+#define QUDA_ARNOLDI 3           // Arnoldi solver
+#define QUDA_IMP_RST_ARNOLDI 4   // Implicitly restarted Arnoldi solver
```
Comment (maddyscientist) [inline include/enum_quda_fortran.h]: Do we need the `SPECTRUM` enums defined in the Fortran interface?

File: include/invert_quda.h [hunk 6]
```diff
@@ -964,17 +1055,55 @@ namespace quda {
...
+public:
+    PreconditionedSolver(Solver &solver, const Dirac &dirac, SolverParam &param, TimeProfile &profile,
+                         const char *prefix) :
+      Solver(param, profile),
+      solver(&solver),
+      dirac(dirac),
+      prefix(prefix)
+    {
+
+      // DMH
```
Comment (maddyscientist) [inline include/invert_quda.h]: Big commented out block here?

File: include/invert_quda.h [hunk 7]
```diff
@@ -1139,16 +1267,22 @@ namespace quda {
...
 
-  public:
-
+public:
     IncEigCG(DiracMatrix &mat, DiracMatrix &matSloppy, DiracMatrix &matPrecon, SolverParam &param, TimeProfile &profile);
 
     virtual ~IncEigCG();
 
+    /**
+       @breif Expands deflation space.
```
Comment (maddyscientist) [inline include/invert_quda.h]: `brief`

File: include/multigrid.h [hunk 8]
```diff
@@ -320,7 +303,7 @@ namespace quda {
     /**
        @brief Dump the null-space vectors to disk.  Will recurse dumping all levels.
     */
-    void dumpNullVectors() const;
+    void dumpNullVectors();
```
Comment (maddyscientist) [inline include/multigrid.h]: Why did this routine lose its `const` qualifier?

File: include/multigrid.h [hunk 9]
```diff
@@ -362,17 +345,34 @@ namespace quda {
...
+       @brief Load the null space vectors in from file
+       @param B Loaded null-space vectors (pre-allocated)
+    */
+    void loadVectors(std::vector<ColorSpinorField *> &B);
+
+    /**
+       @brief Save the null space vectors in from file
+       @param B Save null-space vectors from here
+    */
+    void saveVectors(const std::vector<ColorSpinorField *> &B);
```
Comment (maddyscientist) [inline include/multigrid.h]: Lost `const` qualifier as well

File: include/quda_arpack_interface.h [hunk 10]
```diff
@@ -7,10 +7,7 @@
 #include <color_spinor_field.h>
 #include <dirac_quda.h>
 #include <vector>
-
-//#ifdef PRIMME_LIB
-//#include "primme.h"
-//#endif
+#include <algorithm>
```
Comment (maddyscientist) [inline include/quda_arpack_interface.h]: Can you move these `std` headers to where they are needed rather than blanket inclusion in this header?

File: lib/check_params.h [hunk 11]
```diff
@@ -452,14 +475,15 @@ void printQudaInvertParam(QudaInvertParam *param) {
   }
 #endif
 
-
 #ifdef INIT_PARAM
   P(use_init_guess, QUDA_USE_INIT_GUESS_NO); //set the default to no
-  //P(compute_null_vector, QUDA_COMPUTE_NULL_VECTOR_NO); //set the default to no
+  // P(deflated_guess, QUDA_DEFLATED_GUESS_NO); //set the default to no
```
Comment (maddyscientist) [inline lib/check_params.h]: any reason these are not deleted since they are commented out?

File: include/ritz_quda.h [hunk 12]
```diff
@@ -10,6 +10,8 @@
 #include <dirac_quda.h>
 #include <typeinfo>
 
+#if 0
```
Comment (maddyscientist) [inline include/ritz_quda.h]: Looks like this file can be deleted

File: CMakeLists.txt (hunk 1 above)
Comment (cpviolator) [reply CMakeLists.txt]: In the old code, if QMP or MPI was deleted, only `libparpack.a` would be linked. ARPACK needs both `libparpack.a` and `libarpack.a` to work with MPI, so we always link with `libarpack.a`, and additionally link with `libparpack.a` for parallel builds.

File: CMakeLists.txt (hunk 2 above)
Comment (cpviolator) [reply CMakeLists.txt]: Probably doesn't need it. I'll test it out without.

File: include/quda.h [hunk 13]
```diff
@@ -814,8 +870,28 @@ extern "C" {
...
+
+  /**
+   * Perform the eigensolve using ARPACK. The problem matrix is defined by the
+   * invert param, the mode of solution is specified by the eig param.
+   * It is assumed that the gauge field has already been loaded via loadGaugeQuda().
+   * @param h_els  Host side eigenvalues
+   * @param h_evs  Host side eigenvectors
+   * @param param  Contains all metadata regarding the type of solve.
+   */
+  void eigensolveARPACK(void *h_evals, void *h_evecs, QudaEigParam *param);
```
Comment (maddyscientist) [inline include/quda.h]: If this is a public function to be called from applications then this should be `eigensolveARPACKQuda` to follow the usual naming convention.  Perhaps it should just be subsumed into `eigensolveQuda` though as an option to use `ARPACK` instead of native QUDA?

File: lib/solver.cpp [hunk 14]
```diff
@@ -239,6 +240,117 @@ namespace quda {
     }
   }
 
+  /*
```
Comment (maddyscientist) [inline lib/solver.cpp]: what's the point of this (since it's commented out)?

File: lib/inv_eigcg_quda.cpp [hunk 15]
```diff
@@ -21,189 +21,209 @@
...
-//special types needed for compatibility with QUDA blas:
-   using RowMajorDenseMatrix = Matrix<Complex, Dynamic, Dynamic, RowMajor>;
-
-   static int max_eigcg_cycles = 4;//how many eigcg cycles do we allow?
-
-
-   enum  class libtype {eigen_lib, magma_lib, lapack_lib, mkl_lib};
-
-   class EigCGArgs{
+  using namespace blas;
```
Comment (maddyscientist) [inline lib/inv_eigcg_quda.cpp]: This file, inv_eigcg_quda.cpp has lots of changes here, but this pull isn't really focused on EigCG.  What is being changed in the algorithm here?

File: lib/eigensolve_quda.cpp [hunk 16]
```diff
@@ -0,0 +1,1510 @@
...
+      sum += s;
+      blas::caxpy(-s, *vecs[i], *rvec[0]);
+    }
+    time_ += clock();
+    time_mb += time_;
+    return sum;
+  }
+
+  // Orthogonalise r against V_[j]
+  Complex EigenSolver::blockOrthogonalise(std::vector<ColorSpinorField *> vecs, std::vector<ColorSpinorField *> rvec,
```
Comment (maddyscientist) [inline lib/eigensolve_quda.cpp]: As much as it pains me, probably should be consistent in naming with other uses of orthogonalize in the library

File: lib/eigensolve_quda.cpp [hunk 17]
```diff
@@ -0,0 +1,1510 @@
...
+
+    // 2. Perform block caxpy: V_i * (L_i)^{-1} * A_i
+    for (int i = 0; i < n_defl; i++) { s[i] /= evals[i].real(); }
+
+    // 3. Accumulate sum vec_defl = Sum_i V_i * (L_i)^{-1} * A_i
+    blas::zero(*vec_defl[0]);
+    blas::caxpy(s, eig_vecs_ptr, vec_defl);
+
+    // Orthonormality check in deflation.
+    for (int i = 0; i < n_defl; i++)
```
Comment (maddyscientist) [inline lib/eigensolve_quda.cpp]: Looks like this can be deleted?

File: lib/eigensolve_quda.cpp [hunk 18]
```diff
@@ -0,0 +1,1510 @@
...
+    // Revert to MdagM (or MMdag) mat vec
+    eig_param->use_norm_op = QUDA_BOOLEAN_YES;
+
+    delete tmp[0];
+  }
+
+  // ARPACK INTERAFCE ROUTINES
+  //--------------------------------------------------------------------------
+
+#ifdef ARPACK_LIB
```
Comment (maddyscientist) [inline lib/eigensolve_quda.cpp]: What's the merit in having the ARPACK stuff in here as opposed to in its own file as before?

File: include/enum_quda_fortran.h (hunk 5 above)
Comment (cpviolator) [reply include/enum_quda_fortran.h]: I don't know, I just put them there because the the old enums were there. Should they go?

File: lib/solver.cpp (hunk 14 above)
Comment (cpviolator) [reply lib/solver.cpp]: This was an attempt to move the vector IO to the solver class. I moved them to the Eigensolver class instead.

File: lib/inv_eigcg_quda.cpp (hunk 15 above)
Comment (cpviolator) [reply lib/inv_eigcg_quda.cpp]: I think I made some attempts and integrating the file, and clang formatting changed a lot of stuff. I'm going to checkout the develop version.

File: include/enum_quda_fortran.h (hunk 5 above)
Comment (maddyscientist) [reply include/enum_quda_fortran.h]: For now I would suggest putting the `QUDA_SPECTRUM_XXX` defines in the enum_fortran_fortan.h file as well, for consistency if nothing else.  Eventual plan is to auto generate the `enum_quda_fortran.h` file from the the `enum_quda.h` one once I figure a good way to do that....

**Comment by mathiaswagner** [comment]
@Jenkins test this please

**Comment by maddyscientist** [comment]
@weinbe2 before we merge this, do you want to give it a once over?  I've tested this fairly broadly in the context of prolongation and coarse-grid deflation.

File: include/multigrid.h [hunk 19]
```diff
@@ -445,7 +442,7 @@ namespace quda {
 
   /**
      @brief Calculate preconditioned coarse links and coarse clover inverse field
-     @param Yhat[out] Preconditioned coarse link field
+     @param Yhat[out] Preconidtioned coarse link field
```
Comment (weinbe2) [inline include/multigrid.h]: Typo! (I'm helping, right?)

File: lib/eigensolve_quda.cpp [hunk 20]
```diff
@@ -0,0 +1,948 @@
...
+    // b_j = ||r||
+    beta[j] = sqrt(blas::norm2(*r[0]));
+
+    // Prepare next step.
+    // v_{j+1} = r / b_j
+    blas::zero(*v[j + 1]);
+    blas::axpy(1.0 / beta[j], *r[0], *v[j + 1]);
+  }
+
+  void TRLM::reorder(std::vector<ColorSpinorField *> &kSpace)
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: Not that I'd hold up the merge with this, but can we accomplish this with one of the standard C++ algorithms?

Comment (maddyscientist) [reply lib/eigensolve_quda.cpp]: What I think is needed here is to create a `std::pair` of eigenvalues and eigenvectors and then do `std::sort` on the eigenvalues.  As noted, not something to delay the pull on.

File: include/multigrid.h (hunk 19 above)
Comment (maddyscientist) [reply include/multigrid.h]: Pushed a fix.

File: lib/eigensolve_quda.cpp [hunk 21]
```diff
@@ -0,0 +1,948 @@
...
+    residua = (double*)safe_malloc(nKr * sizeof(double));
+    for (int i = 0; i < nKr; i++) { residua[nKr] = 0.0; }
+
+    // Quda MultiBLAS freindly array
+    Qmat = (Complex*)safe_malloc(nEv * nKr * sizeof(Complex));
+
+    // Part of the spectrum to be computed.
+    spectrum = strdup("SR"); // Initialised to stop the compiler warning.
+
+    if (eig_param->use_poly_acc) {
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: In the case of polynomial acceleration: do we want to make anything that's not `SR` throw an error? (except maybe `LR`, which may be worth a warning instead---it's highly suboptimal but it's not really wrong if I remember correctly). 

There's a big asterisk here in that iirc Chebyshev acceleration can be proven to work for spectra that aren't purely real as long as they lie in a well-behaved ellipse in the right half plane (i.e., I believe you can get it to work on a Wilson-clover spectrum with enough experimenting), so that'd motivate just letting this be for now...

File: lib/eigensolve_quda.cpp [hunk 22]
```diff
@@ -0,0 +1,948 @@
...
+  // Orthogonalise r against V_[j]
+  Complex EigenSolver::orthogonalize(std::vector<ColorSpinorField *> vecs, std::vector<ColorSpinorField *> rvec, int j)
+  {
+    time_ = -clock();
+    Complex s(0.0, 0.0);
+    Complex sum(0.0, 0.0);
+    for (int i = 0; i < j; i++) {
+      s = blas::cDotProduct(*vecs[i], *rvec[0]);
+      sum += s;
+      blas::caxpy(-s, *vecs[i], *rvec[0]);
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: Can we plug in the hybrid modified Gram-Schmidt/Gram-Schmidt ortho here?

File: lib/eigensolve_quda.cpp [hunk 23]
```diff
@@ -0,0 +1,948 @@
...
+      sigma_old = sigma;
+    }
+    blas::copy(out,*tmp2);
+
+    delete tmp1;
+    delete tmp2;
+  }
+
+  // Orthogonalise r against V_[j]
+  Complex EigenSolver::orthogonalize(std::vector<ColorSpinorField *> vecs, std::vector<ColorSpinorField *> rvec, int j)
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: Is there any reason we have this around since we have `blockOrthogonalize`?

File: lib/eigensolve_quda.cpp [hunk 24]
```diff
@@ -0,0 +1,948 @@
...
+
+    return;
+  }
+
+  void EigenSolver::loadFromFile(const DiracMatrix &mat, std::vector<ColorSpinorField *> &kSpace,
+                                 std::vector<Complex> &evals)
+  {
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Loading eigenvectors\n");
+    loadVectors(kSpace, eig_param->vec_infile);
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: This comment would be better put in `qio_field.cpp`, but anyway: I think we should have a more constructive error when a file doesn't exist, especially that loading/saving eigenvectors is becoming a lot more important in QUDA with all these additions. Right now the only error message in `qio_field.cpp` is that "Open file failed!", and I think we should say something more useful (if QIO gives us that info), i.e., if we can discriminate "file doesn't exist" from "file isn't properly formatted", that's a little thing that goes a long way.

File: lib/eigensolve_quda.cpp [hunk 25]
```diff
@@ -0,0 +1,948 @@
...
+    kSpace.push_back(ColorSpinorField::Create(csParam));
+    csParam.create = QUDA_ZERO_FIELD_CREATE;
+    r.push_back(ColorSpinorField::Create(csParam));
+    //---------------------------------------------------------------------------
+
+    // Convergence and locking criteria
+    double mat_norm = 0.0;
+    double epsilon = DBL_EPSILON;
+    QudaPrecision prec = kSpace[0]->Precision();
+    if (prec == QUDA_DOUBLE_PRECISION) {
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: We should turn this `if` chain into ` switch` statement, and further, we should add `#ifdef` guards that respect the precision `cmake` variable (and throw an error if it's an invalid precision).

File: lib/eigensolve_quda.cpp [hunk 26]
```diff
@@ -0,0 +1,948 @@
...
+        kSpace.push_back(ColorSpinorField::Create(csParam));
+      }
+    }
+
+    for (int i = 0; i < iter_keep; i++) {
+      int k = offset + i;
+      *r[0] = *kSpace[num_locked];
+      blas::ax(ritz_mat[dim * i], *r[0]);
+      *kSpace[k] = *r[0];
+      for (int j = 1; j < dim; j++) { blas::axpy(ritz_mat[i * dim + j], *kSpace[num_locked + j], *kSpace[k]); }
```
Comment (weinbe2) [inline lib/eigensolve_quda.cpp]: Not that I want to add more multi-blas routines, but for performance reasons would it ultimately be a net win to replace this with a batch `caxpy` as opposed to multiple `axpy`s? (Bandwidth savings compensating for unnecessary arithmetic costs)

Comment (weinbe2) [reply lib/eigensolve_quda.cpp]: At least be a net win*

File: lib/interface_quda.cpp [hunk 27]
```diff
@@ -2403,155 +2404,137 @@ void cloverQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
...
-
-  // download source
-  ColorSpinorParam cudaParam(cpuParam, *param);
-  cudaParam.create = QUDA_COPY_FIELD_CREATE;
-  r = new cudaColorSpinorField(*h_r, cudaParam);
-  Apsi = new cudaColorSpinorField(*h_Apsi, cudaParam);
+  ColorSpinorParam cudaParam(cpuParam);
+  cudaParam.location = QUDA_CUDA_FIELD_LOCATION;
+  cudaParam.create = QUDA_ZERO_FIELD_CREATE;
+  eig_param->cuda_prec_ritz == QUDA_DOUBLE_PRECISION ? cudaParam.fieldOrder = QUDA_FLOAT2_FIELD_ORDER :
```
Comment (weinbe2) [inline lib/interface_quda.cpp]: Is this strictly correct? Isn't it always `FLOAT2` for staggered fields in general?

File: lib/inv_cg_quda.cpp [hunk 28]
```diff
@@ -310,10 +346,30 @@ namespace quda {
     // compute initial residual
     double r2 = 0.0;
     if (param.use_init_guess == QUDA_USE_INIT_GUESS_YES) {
-      mat(r, x, y, tmp3);
-      r2 = blas::xmyNorm(b, r);
-      if (b2 == 0) b2 = r2;
-      blas::copy(y, x);
+
+      // Just replace any initial guess with a deflated RHS
```
Comment (weinbe2) [inline lib/inv_cg_quda.cpp]: This is a bad idea: I've seen cases in MILC where sometimes CG is called to "polish" a solution (i.e., check and make sure it's a solution, and if it's not polish it up so it is, which is normally just a few iterations). In cases like that, completely nuking the existing initial guess is suboptimal. We should work out the logic to avoid this problem.

Comment (weinbe2) [reply lib/inv_cg_quda.cpp]: That said, I don't know if there's an intersection of deflation and polishing solutions in practice in general, but still, this shouldn't be too difficult to work out.

File: lib/multigrid.cpp [hunk 29]
```diff
@@ -894,9 +968,8 @@ namespace quda
     } else { // do the coarse grid solve
 
       ColorSpinorField *out=nullptr, *in=nullptr;
-
+      // Deflate EVEN ODD here
```
Comment (weinbe2) [inline lib/multigrid.cpp]: Is this comment still relevant?

File: lib/multigrid.cpp [hunk 30]
```diff
@@ -1429,4 +1387,53 @@ namespace quda
...
+    // This is the vector precision used by matResidual
+    csParam.setPrecision(param.mg_global.invert_param->cuda_prec_sloppy, QUDA_INVALID_PRECISION, true);
+
+    for (int i = 0; i < nKr; i++) B_evecs.push_back(ColorSpinorField::Create(csParam));
+
+    // before entering the eigen solver, lets free the B vectors to save some memory
+    ColorSpinorParam bParam(*param.B[0]);
+    for (int i = 0; i < (int)param.B.size(); i++) delete param.B[i];
+
+    DiracMdagM *mdagm = new DiracMdagM(*diracResidual);
```
Comment (weinbe2) [inline lib/multigrid.cpp]: Do we want to expose the option of using `DiracMMdag` instead? (In practice I don't think it makes sense to use left singular vectors, so this isn't a hard request.)

File: lib/quda_arpack_interface.cpp [hunk 31]
```diff
@@ -1,384 +1,541 @@
...
+    // Construct parameters and memory allocation
+    //---------------------------------------------------------------------------------
+    double time_ar = 0.0; // time in ARPACK
+    double time_mv = 0.0; // time in QUDA mat vec + data transfer
+    double time_ev = 0.0; // time in computing Eigenvectors
+
+    // MPI objects
+    int *fcomm_ = nullptr;
+#if (defined(QMP_COMMS) || defined(MPI_COMMS))
+    MPI_Fint mpi_comm_fort = MPI_Comm_c2f(MPI_COMM_WORLD);
```
Comment (weinbe2) [inline lib/quda_arpack_interface.cpp]: This needs to be `MPI_COMM_HANDLE` w/regards to the multiple communicator support.

Comment (weinbe2) [reply lib/quda_arpack_interface.cpp]: In lieu of adding this comment anywhere else (without me checking if it actually does appear anywhere else), just run a `grep -r "MPI_COMM_WORLD"` and make sure it doesn't show up anywhere except for setting the default value of `MPI_COMM_HANDLE` :)

File: lib/quda_arpack_interface.cpp [hunk 32]
```diff
@@ -1,384 +1,541 @@
...
+    iparam_[6] = 1;
+
+    // ARPACK problem type to be solved
+    char howmny = 'P';
+    char bmat = 'I';
+    char *spectrum;
+    spectrum = strdup("SR"); // Initialsed just to stop the compiler warning...
+
+    if (eig_param->use_poly_acc) {
+      if (eig_param->spectrum == QUDA_SPECTRUM_SR_EIG)
```
Comment (weinbe2) [inline lib/quda_arpack_interface.cpp]: See comment I've put elsewhere about throwing warnings/errors about polynomial acceleration for non-normal ops

File: lib/eigensolve_quda.cpp (hunk 20 above)
Comment (weinbe2) [reply lib/eigensolve_quda.cpp]: Update: it looks like this has already been worked out in `lib/quda_arpack_interface.cpp`, let's grab it from there.

File: tests/test_util.cpp [hunk 33]
```diff
@@ -1799,11 +1846,21 @@ void usage(char** argv )
   printf("    --mass-normalization                      # Mass normalization (kappa (default) / mass / asym-mass)\n");
   printf("    --matpc                                   # Matrix preconditioning type (even-even, odd-odd, even-even-asym, odd-odd-asym) \n");
   printf("    --solve-type                              # The type of solve to do (direct, direct-pc, normop, normop-pc, normerr, normerr-pc) \n");
+  printf("    --solution-type                           # The solution we desire (mat (default), mat-dag-mat, mat-pc, "
+         "mat-pc-dag-mat-pc (default for multi-shift))\n");
```
Comment (weinbe2) [inline tests/test_util.cpp]: Except for staggered!

Comment (weinbe2) [reply tests/test_util.cpp]: iirc The `--solve-type` is `direct-pc` and the `--solution-type` is `mat-pc`, has anyone made sure this isn't broken by the default changes? I _think_ `staggered_invert_test` hardcodes a bunch of these things if you request the multishift tests.

File: CMakeLists.txt [hunk 34]
```diff
@@ -326,9 +326,14 @@ if(QUDA_MPI OR QUDA_QMP)
     message(STATUS "Found environment variable MPICC. Using it for MPI detection: $ENV{MPICC}")
     set(MPI_C_COMPILER $ENV{MPICC})
     set(mpimessage True)
+  endif()
+  if((NOT MPI_FORT_COMPILER) AND DEFINED ENV{MPIFORT})
+    message(STATUS "Found environment variable MPIFORT. Using it for MPI detection: $ENV{MPIFORT}")
+    set(MPI_C_COMPILER $ENV{MPICC})
```
Comment (mathiaswagner) [inline CMakeLists.txt]: This is not what we want (MPI_C_COMPILER) - but I am thinking about dropping support for that anyway.

File: CMakeLists.txt [hunk 35]
```diff
@@ -411,28 +416,36 @@ endif(QUDA_MAGMA)
 
 # This selects arpack or parpack for Multi GPU
 if(QUDA_ARPACK)
+  enable_language(Fortran)
```
Comment (mathiaswagner) [inline CMakeLists.txt]: duplicated. But I think for FindMPI to work properly this needs to come before FindMPI anyway

File: lib/eigensolve_quda.cpp (hunk 25 above)
Comment (maddyscientist) [reply lib/eigensolve_quda.cpp]: I'll fix this in my next push.  However, I don't agree with respect to cmake variable since this code doesn't involve template instantiation.

File: lib/eigensolve_quda.cpp (hunk 26 above)
Comment (maddyscientist) [reply lib/eigensolve_quda.cpp]: In the `axpy`, can the indices `k` and `num_locked + j` be equal?  Just wondering hos easy this is multi-blas-ify

File: lib/interface_quda.cpp (hunk 27 above)
Comment (maddyscientist) [reply lib/interface_quda.cpp]: Fixed in next push

File: lib/quda_arpack_interface.cpp (hunk 31 above)
Comment (maddyscientist) [reply lib/quda_arpack_interface.cpp]: Fixed in next push

File: tests/test_util.cpp (hunk 33 above)
Comment (maddyscientist) [reply tests/test_util.cpp]: Yes, staggered hard codes a bunch of things, and has done for ages (outside of this pull).  For reasons I can't recall, staggered_invert_test uses `--test` to set these things.  This does need to be fixed for staggered to make it consistent.  Not on this pull.

File: lib/inv_cg_quda.cpp (hunk 28 above)
Comment (maddyscientist) [reply lib/inv_cg_quda.cpp]: I agree, this should be fixed.  It would also come if we were at some point to combine deflation with multi-shift solvers, e.g., where we run deflated staggered multi shift in mixed precision and then want to refine using deflated CG.

File: lib/eigensolve_quda.cpp (hunk 26 above)
Comment (cpviolator) [reply lib/eigensolve_quda.cpp]: This can multi-blas-ified. `num_locked + j` is never greater than `dim`, which is never greater than `nKr`. `k` is always equal to `nKr + 1`, so the vectors never alias.

File: lib/eigensolve_quda.cpp (hunk 21 above)
Comment (cpviolator) [reply lib/eigensolve_quda.cpp]: Ok, leaving this alone for the moment.

File: lib/eigensolve_quda.cpp (hunk 23 above)
Comment (cpviolator) [reply lib/eigensolve_quda.cpp]: Nope, it has been banished :)

File: lib/eigensolve_quda.cpp (hunk 24 above)
Comment (cpviolator) [reply lib/eigensolve_quda.cpp]: Is this a general comment about how our QIO code handles errors, or about how I use it in the eigensolver? Shall we fix the QIO `stdout` on this pull?

File: lib/multigrid.cpp (hunk 30 above)
Comment (cpviolator) [reply lib/multigrid.cpp]: If it doesn't make sense, then I won't expose the option. We use Lanczos here not to extract any particular information about the spectrum of a particular operator (as we would in the stand alone solver) but to simply deflate a CG solve.

File: lib/quda_arpack_interface.cpp (hunk 32 above)
Comment (cpviolator) [reply lib/quda_arpack_interface.cpp]: Again, leaving this alone. For normal operators checks, see here: https://github.com/lattice/quda/blob/feature/deflatedMG/lib/interface_quda.cpp#L2470

File: lib/multigrid.cpp (hunk 30 above)
Comment (maddyscientist) [reply lib/multigrid.cpp]: It does make sense to expose the option.  When doing CGNE, you solve using MMdag not MMdag.

File: lib/inv_cg_quda.cpp (hunk 28 above)
Comment (maddyscientist) [reply lib/inv_cg_quda.cpp]: You form new residual equation, e.g., `r = b - Ax_0`, which is what you do normally with an initial guess.  Now the system you deflate is `A y = r`, and the total solution vector is given by `x = x0 + y`.

File: include/invert_quda.h [hunk 36]
```diff
@@ -1139,16 +1236,22 @@ namespace quda {
...
-  public:
-
+public:
     IncEigCG(DiracMatrix &mat, DiracMatrix &matSloppy, DiracMatrix &matPrecon, SolverParam &param, TimeProfile &profile);
 
     virtual ~IncEigCG();
 
+    /**
+       @brief Expands deflation space.
+       @param V container of new eigenvectors
```
Comment (mathiaswagner) [inline include/invert_quda.h]: How is V a container? Is it using a Composite field?

File: lib/deflation.cpp [hunk 37]
```diff
@@ -125,33 +125,34 @@ namespace quda {
     std::vector<ColorSpinorField*> res;
     res.push_back(r);
 
-    for(int i = 0; i < nevs_to_print; i++)
-    {
-       zero(*r);
+    for (int i = 0; i < nevs_to_print; i++) {
+      zero(*r);
```
Comment (mathiaswagner) [inline lib/deflation.cpp]: not new in this pull but the blank lines confused me.

**Comment by maddyscientist** [comment]
The addition of a staggered eigenvector is set to #821 and will for a later pull.

File: lib/eigensolve_quda.cpp (hunk 24 above)
Comment (maddyscientist) [reply lib/eigensolve_quda.cpp]: Ok, while that specific print statement doesn't belong there since there is another equivalent statement in `EigenSolver::loadVectors` which called from this function (and so I'll remove it on my next push), I do see that the info from QIO is actually not bad.  E.g., if QIO fails to find a file, it will print the following
```
QIO_create_reader(0): cannot open conf_stout_1e4.1000__level_0_nvec_6 as SINGLEFILE; trying PARTFILE
QIO_create_reader(0): cannot open conf_stout_1e4.1000__level_0_nvec_6.vol0000 as PARTFILE; trying PARTFILE_DIR
QIO_create_reader(0): cannot open vol0000/conf_stout_1e4.1000__level_0_nvec_6 as PARTFILE_DIR; QIO_create_reader FAILED
MG level 0 (GPU): open_test_input(0): QIO_open_read returns NULL.
MG level 0 (GPU): ERROR: MG level 0 (GPU): ERROR: Open file failed
```
This seems plenty informative.

File: lib/deflation.cpp (hunk 37 above)
Comment (maddyscientist) [reply lib/deflation.cpp]: Fixed in my next push

File: include/invert_quda.h (hunk 36 above)
Comment (maddyscientist) [reply include/invert_quda.h]: Yes, this is a composite field.  Made this clear in my next push.

File: lib/quda_arpack_interface.cpp (hunk 32 above)
Comment (maddyscientist) [reply lib/quda_arpack_interface.cpp]: Marking this as resolved...

File: CMakeLists.txt (hunk 34 above)
Comment (maddyscientist) [reply CMakeLists.txt]: Fixed in #820

File: CMakeLists.txt (hunk 35 above)
Comment (maddyscientist) [reply CMakeLists.txt]: Fixed in #820

File: lib/multigrid.cpp (hunk 30 above)
Comment (maddyscientist) [reply lib/multigrid.cpp]: Noting this is fixed now

File: lib/interface_quda.cpp [hunk 38]
```diff
@@ -2403,155 +2404,157 @@ void cloverQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
...
-  if (getVerbosity() >= QUDA_VERBOSE) {
-    cpu = blas::norm2(*h_r);
-    gpu = blas::norm2(*r);
-    printfQuda("r vector CPU %1.14e CUDA %1.14e\n", cpu, gpu);
-    cpu = blas::norm2(*h_Apsi);
-    gpu = blas::norm2(*Apsi);
-    printfQuda("Apsi vector CPU %1.14e CUDA %1.14e\n", cpu, gpu);
+  // If you use polynomial acceleration on a non-symmetric matrix,
+  // the solver will fail.
+  if (eig_param->use_poly_acc && !eig_param->use_norm_op && !(inv_param->dslash_type == QUDA_LAPLACE_DSLASH)) {
```
Comment (weinbe2) [inline lib/interface_quda.cpp]: We need a special check for staggered/asqtad PC operators, since they're normal.

Comment (weinbe2) [reply lib/interface_quda.cpp]: ^and therefore safe

**Comment by maddyscientist** [comment]
@cpviolator and I have done some further cleanup and more testing.  Given approval from @weinbe2 already, we're ready to merge.
